clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h

```diff
@@ -373,34 +374,117 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
 
   exchange(uint8_t *local_memory) : _local_memory(local_memory) {}
 
-  /// Rearrange elements from rank order to blocked order
-  template <typename Item>
-  __dpct_inline__ void
-  scatter_to_blocked(Item item, T (&keys)[VALUES_PER_THREAD],
-                     int (&ranks)[VALUES_PER_THREAD]) {
+  static size_t adjust_by_padding(size_t offset) {
+
+    if constexpr (INSERT_PADDING) {
+      offset = detail::shr_add(offset, LOG_LOCAL_MEMORY_BANKS, offset);
+    }
+    return offset;
+  }
+
+  template <typename Item> struct getStripedOffset {
+    size_t operator()(Item item, int i) {
+      size_t offset =
+          size_t(item.get_local_id(0) * VALUES_PER_THREAD) + size_t(i);
+      return adjust_by_padding(offset);
+    }
+  };
+
+  template <typename Item> struct getBlockedOffset {
+    size_t operator()(Item item, int i) {
+      size_t offset =
+          i * item.get_local_range(2) * item.get_local_range(1) *
+                 item.get_local_range(0) +
+          item.get_local_id(0);
+      return adjust_by_padding(offset);
+    }
+  };
+
+  template <typename Item, typename Iterator> struct getScatterOffset {
+    Iterator begin;
+    Iterator end;
+    getScatterOffset(const int (&ranks)[VALUES_PER_THREAD]) {
+      begin = ranks;
+      end = ranks + VALUES_PER_THREAD;
+    }
+    ~getScatterOffset() {}
+    size_t operator()(Item item, int i) const {
+      Iterator it = begin + i;
+      size_t offset;
+      if (it >= begin && it < end) {
+        offset = *it;
+      }
+      return adjust_by_padding(offset);
+    }
```

Comment:
```suggestion
    size_t operator()(Item item, int i) const {
      return adjust_by_padding(begin[i]);
    }
```
I think its reasonable to assume in this setting that begin is a random access iterator, and also that `i` is within bounds of `VALUES_PER_THREAD`.  You can add some comments here stating assumptions, but these checks will slow things down in a low level helper and not actually prevent any errors.
If you take this suggestion, you can also remove the `end` iterator.

Reply:
Modified.